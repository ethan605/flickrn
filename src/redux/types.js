export const PREFIXES = {
  FLICKR: 'flickr/',
  SHARED: 'shared/',
};

export const FLICKR = {
  CLEAR_PHOTO_INFO: 'flickr/CLEAR_PHOTO_INFO',
  GET_PHOTO_INFO: 'flickr/GET_PHOTO_INFO',
  SEARCH_PHOTOS: 'flickr/SEARCH_PHOTOS',
};

export const SHARED = {
};
