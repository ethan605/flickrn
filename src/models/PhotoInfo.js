import _ from 'lodash';
import Joi from 'react-native-joi';

// Locals
import BaseModel, { COMMON_SCHEMAS } from './BaseModel';

const { INTEGER, OBJECT, STRING } = COMMON_SCHEMAS;

export default class PhotoInfo extends BaseModel {
  static dataValidator(): Joi.SchemaMap {
    return {
      description: OBJECT,
      farm: INTEGER.required(),
      id: STRING.required(),
      originalformat: STRING,
      originalsecret: STRING,
      owner: OBJECT.required(),
      secret: STRING.required(),
      server: STRING.required(),
      title: OBJECT.required(),
    };
  }

  get titleContent() {
    const { _content: content } = this.title || {};
    return content;
  }

  get descriptionContent() {
    const { _content: content } = this.description || {};
    return content;
  }

  get originalImage() {
    const { farm, id, originalformat, originalsecret, secret, server } = this;

    if (!_.isEmpty(originalsecret) && !_.isEmpty(originalformat))
      return `https://farm${farm}.staticflickr.com/${server}/${id}_${originalsecret}_o.${originalformat}`;

    return `https://farm${farm}.staticflickr.com/${server}/${id}_${secret}_b.jpg`;
  }
}
