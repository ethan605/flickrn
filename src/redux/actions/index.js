import * as flickr from './flickr';

/* eslint-disable import/prefer-default-export */

export { flickr };
