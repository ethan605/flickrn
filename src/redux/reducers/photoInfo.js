import _ from 'lodash';
import { FLICKR } from 'src/redux/types';

// Constants
import { AXIOS_REQUEST_SUFFIXES } from 'src/constants';

// Models
import PhotoInfo from 'src/models/PhotoInfo';

const INITIAL_STATE = {
  data: undefined,
  errorMessage: undefined,
  loading: false,
};

const { ERROR, SUCCESS } = AXIOS_REQUEST_SUFFIXES;

export default function photoInfoReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === FLICKR.CLEAR_PHOTO_INFO)
    return INITIAL_STATE;

  if (type === FLICKR.GET_PHOTO_INFO + ERROR) {
    const { error: { message: errorMessage } } = action;
    return { ...state, errorMessage, loading: false };
  }

  if (type === FLICKR.GET_PHOTO_INFO + SUCCESS) {
    const { payload: { photo } } = action;
    if (_.isEmpty(photo)) return state;

    return { ...state, data: PhotoInfo.build(photo), loading: false };
  }

  if (type === FLICKR.GET_PHOTO_INFO)
    return { ...state, data: undefined, errorMessage: undefined, loading: true };

  return state;
}
