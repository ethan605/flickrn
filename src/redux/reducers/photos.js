import _ from 'lodash';
import { FLICKR } from 'src/redux/types';

// Constants
import { AXIOS_REQUEST_SUFFIXES } from 'src/constants';

// Models
import Photo from 'src/models/Photo';

const INITIAL_STATE = {
  currentPage: 1,
  data: [],
  errorMessage: undefined,
  hasMore: true,
  loading: false,
};

const { ERROR, SUCCESS } = AXIOS_REQUEST_SUFFIXES;

export default function photosReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === FLICKR.SEARCH_PHOTOS + ERROR) {
    const { error: { message: errorMessage } } = action;
    return { ...state, errorMessage, hasMore: false, loading: false };
  }

  if (type === FLICKR.SEARCH_PHOTOS + SUCCESS) {
    const { payload: { photos } } = action;
    if (_.isEmpty(photos)) return state;

    const { page, pages, photo } = photos;
    const hasMore = page < pages;

    const { data: currentPhotos } = state;
    const newPhotos = Photo.buildArray(photo);

    const data = page === 1 ? newPhotos : [...currentPhotos, ...newPhotos];
    return { ...state, hasMore, data, currentPage: page, loading: false };
  }

  if (type === FLICKR.SEARCH_PHOTOS)
    return { ...state, errorMessage: undefined, loading: true };

  return state;
}
