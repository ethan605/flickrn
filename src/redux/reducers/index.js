import photoInfo from './photoInfo';
import photos from './photos';

export { photoInfo, photos };
