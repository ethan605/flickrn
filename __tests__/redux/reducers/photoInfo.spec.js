import { photoInfo as photoInfoReducer } from 'src/redux/reducers';
import { FLICKR } from 'src/redux/types';

// Helpers
import buildMockState from '__tests__/helpers/buildMockState';
import { buildAxiosActions } from '__tests__/helpers/mockApiRequests';

// Models
import PhotoInfo from 'src/models/PhotoInfo';

describe('redux/reducers/photoInfo', () => {
  const { photoInfo: mockState } = buildMockState();
  const mockSuccess = require('__tests__/fixtures/getPhotoInfo/success.json');
  const mockError = require('__tests__/fixtures/getPhotoInfo/error.json');

  const requestAction = { type: FLICKR.GET_PHOTO_INFO };

  const { photo } = mockSuccess;
  const { errorAction, successAction } = buildAxiosActions(requestAction, { photo }, mockError);

  test('start', () => {
    const currentState = {
      ...mockState,
      data: { title: 'some random data' },
      errorMessage: 'some random error',
    };

    const expectedState = {
      ...mockState,
      loading: true,
    };

    const reduced = photoInfoReducer(currentState, requestAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('success', () => {
    const currentState = {
      ...mockState,
      loading: true,
    };

    const expectedState = {
      ...mockState,
      data: PhotoInfo.build(mockSuccess.photo),
    };

    const reduced = photoInfoReducer(currentState, successAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('error', () => {
    const currentState = {
      ...mockState,
      loading: true,
    };

    const expectedState = {
      ...mockState,
      errorMessage: 'Random flickr.photos.getInfo error',
    };

    const reduced = photoInfoReducer(currentState, errorAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('refresh', () => {
    const data = PhotoInfo.build(mockSuccess.photo);
    
    const currentState = { ...mockState, data };
    const expectedState = { ...mockState, loading: true };

    const reduced = photoInfoReducer(currentState, requestAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });
});
