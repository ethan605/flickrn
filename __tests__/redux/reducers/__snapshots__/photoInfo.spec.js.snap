// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/reducers/photoInfo error 1`] = `
Object {
  "data": undefined,
  "errorMessage": "Random flickr.photos.getInfo error",
  "loading": false,
}
`;

exports[`redux/reducers/photoInfo refresh 1`] = `
Object {
  "data": undefined,
  "errorMessage": undefined,
  "loading": true,
}
`;

exports[`redux/reducers/photoInfo start 1`] = `
Object {
  "data": undefined,
  "errorMessage": undefined,
  "loading": true,
}
`;

exports[`redux/reducers/photoInfo success 1`] = `
Object {
  "data": PhotoInfo {
    "description": Object {
      "_content": "Hair : #Tableau Vivant  \\\\\\\\ Silvana hair <a href=\\"https://maps.secondlife.com/secondlife/8 8/108/194/1085\\" rel=\\"nofollow\\">Collabor88</a>

Outfit :   #New in <a href=\\"http://maps.secondlife.com/secondlife/kustom9/124/120/22\\" rel=\\"nofollow\\">kustom9 </a>
                 #=Zenith= Girl Heart Uniforms gacha 
                 =Zenith=Mobile neck case gacha
                 =Zenith=Leather Star Boots with Sock gacha 
    
backpack :  =Zenith=Cat Backpack (all Color) -RARE gacha 

 Cat : #JIAN Chubby Cat Bird Bus

Dogs : #+Half-Deer+  Corgi Puppy - Walking Companion RARE
            +Half-Deer+    Corgi Puppy - Cuddle Companion RARE
",
    },
    "farm": 5,
    "id": "40107995774",
    "originalformat": "png",
    "originalsecret": "0e602e4261",
    "owner": Object {
      "iconfarm": 5,
      "iconserver": "4798",
      "location": "",
      "nsid": "152508136@N06",
      "path_alias": null,
      "realname": "jodi H",
      "username": "jody2015SL",
    },
    "secret": "7ed5905c16",
    "server": "4778",
    "title": Object {
      "_content": "üê±üê∂üê∂ üíñ",
    },
  },
  "errorMessage": undefined,
  "loading": false,
}
`;
